[dagit] Step execution time + recent successes/fails in the graph sidebar

bengotow opened the pull request:
## Summary
This is a bit of an exploratory PR based on our convos this week - I think there are some pros/cons to doing this and I'm curious to get thoughts.

Our new Asset graph explorer allows you to click on individual assets and see step execution time + graphs of materialization event metadata, but there is no corresponding view for Ops.  After mostly working on the asset UI for the last month it feels pretty absent. Users have also noted they have to use asset materializations to get a step execution time graph.

This PR adds a `Solid.stepStats` resolver that pulls the last N step stats for the step with the op's name in the parent pipeline / job. This is technically abusing the fact that `step key == op name` and I'm not sure if there's a great way around that, but we're slightly blurring the distinction elsewhere in Dagit already. (Links from solid to run logs etc.)

The end result is that clicking an op in a job shows you a graph of it's recent step execution time and success / failure. You can hover over the graphs and click the run dots to jump to the step logs in the runs. I think this feels pretty nice and there are some other interesting things we might add:

- "Last invocation logs" modal that lets you view the last step execution logs right in the graph explorer
- Overlays on the DAG visualization similar to the ones in the asset graph showing last success / failure or last execution duration.

One potential downside of this change is that it blends operational data into this definition view. This is already the case in the new asset graph, but moving forward with this would commit us to that path a bit more and I wouldn't want it to overlap with the instance-wide operational view that @braunjj is working on.

Right now this runs a separate GraphQL query in the sidebar with it's own loading spinner because the resolver makes 21 database queries.

<img width="1782" alt="image" src="https://user-images.githubusercontent.com/1037212/147267959-246d33f0-29d0-471e-8fef-065e4d9e53d5.png">


## Test Plan
<!--- Please describe the tests you have added and your testing environment (if applicable). -->




## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask in our Slack. -->

- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

prha reviewed (commented):
How does this work with graphs / composite solids?  My understanding is that composite solids both implement the `Solid` interface as well as the `SolidContainer` interface.  I think that is the main place where the op => step_key conversion doesn't quite make sense.  

If we're only showing this for leaf solids, this makes sense to me, but I would consider asserting that in the resolver / splitting that out using the `__typename` checks in the client.

Putting operational information in the sidebar is pretty interesting.  When no op is selected, would I expect to see overall job status (e.g. last 10 runs) in the right nav?  Or do we jam that up at the top?

bengotow commented:
Ahh that's a good call re: composite solids, we should definitely make this resolver available on basic ops only. Right now I have the UI setup to only display this section of the sidebar when you're viewing a job (vs a graph), but maybe we can tighten that assertion as well.

I agree it'd make sense to have an operational view of some sort without a solid selected - curious to get @braunjj's thoughts about how far we want to go :-)

hellendag commented on js_modules/dagit/packages/core/src/pipelines/SidebarOpExecutionGraphs.tsx:
```diff
@@ -0,0 +1,153 @@
+                    >
+                      <StepStatusDot
+                        onMouseEnter={() => startTime && setHighlightedStartTime(startTime * 1000)}
+                        onMouseOut={() => setHighlightedStartTime(null)}
```
`onMouseLeave`?

hellendag commented on js_modules/dagit/packages/core/src/pipelines/SidebarOpExecutionGraphs.tsx:
```diff
@@ -0,0 +1,153 @@
+      <SidebarSection title={'Execution Status'}>
+        <Box padding={{left: 24, right: 16, vertical: 12}}>
+          <Box flex={{gap: 16}} style={{fontSize: '0.8rem'}}>
+            <div style={{flex: 1}}>{`Last ${Math.min((stats || []).length, 10)} Runs`}</div>
```
Nit, plural won't work for 1 run

bengotow replied:
Ahh good catch!

bengotow commented:
Hey folks! @prha  I addressed your point about the composite solids. The GraphQL for this feature now looks like this:

```
    ... on Pipeline {
      solidHandles{
        handleID
        stepStats(limit: 15) {
          ... on SolidStepStatusUnavailableError {
            message
          }
          ... on SolidStepStatsConnection {
            nodes {
              startTime
            }
          }
        }
      }
    }
```

For composite solids / subgraphs and for dynamic ops, a `SolidStepStatusUnavailableError` error is returned and the UI hides the sidebar section quietly.

I also realized that my solution didn't support ops inside subgraphs (where the step key is `parent_op.child_op`). To fix this I moved the resolver up to the solidHandle, where the handle ID more accurately matches the step name.

Also fixed your comments dish, thanks!

prha commented on python_modules/dagster-graphql/dagster_graphql/schema/solids.py:
```diff
@@ -484,6 +517,28 @@ def __init__(self, handle, solid, parent=None):
+        self._solid = solid
+
+    def resolve_stepStats(self, _graphene_info, limit):
+        if self._solid.get_is_dynamic_mapped():
```
this is pretty good!

prha reviewed (approved):
I think this looks great!
